### ISSUE TITLE
empty() return "wrong" value after view::drop

### ISSUE THREAD

**Issue by tower120** [ericniebler/range-v3#813]
The following code :

```c++
    std::vector<int> list = {1,2,3,4,5};

    auto l = list | view::drop(20);
    bool b = l.empty();
    std::cout << b;
```

output 0

Is this by the design?

**Comment by gnzlbg** [comment]
Yes, because the `View` `l` is not empty yet. To make it empty you have to exhaust it first (e.g. by iterating over it until it becomes empty, although calling `begin` on it should already be enough).

**Comment by tower120** [comment]
From doc:
```
view::drop
    Given a source range and an integral count, return a range consisting of all but the first count elements from the source range, or an empty range if it has fewer elements.
```

>The View l is not empty yet. To make it empty you have to exhaust it first...

How/why it is not empty? *[Your suggestion with begin works]*
When view "considered" empty?

**Comment by gnzlbg** [comment]
For example:

```c++
int main(int /*argc*/, char ** /*argv*/) {
    std::vector<int> list = {1,2,3,4,5};
    auto l = list | ranges::view::drop(3);
    std::cout << l.empty() << std::endl;  // false
    for (auto&& e : l) { std::cout << "printed twice" << std::endl; }
    std::cout << l.empty() << std::endl; // false
    for (auto&& e : l) { std::cout << "printed twice again" << std::endl; }
    std::cout << l.empty() << std::endl; // false
}
```

**Comment by gnzlbg** [comment]
@tower120 views are lazy algorithms, creating a view does no work (or very very little work). When you iterate over a view is when work starts getting done. In this case, begin will be incremented till the first element is found, and in your case the end will be arrived and then the view is "empty".

**Comment by gnzlbg** [comment]
@ericniebler the following produces an infinite loop on wandbox:

```c++
int main(int /*argc*/, char ** /*argv*/) {
    std::vector<int> list = {1,2,3,4,5};
    auto l = list | ranges::view::drop(6);
    for (auto&& e : l) { std::cout << "never printed" << std::endl; }
}
```

**Comment by tower120** [comment]
Ok, why it becomes empty when I drop **exactly** all elements in container?

```c++
    std::vector<int> list = {1,2,3,4,5};

    auto l = list | view::drop(5);
    bool b = l.empty();
    std::cout << b;
```
Output 1.

I do believe, that accordingly to doc it should becomes `empty() == true` right at the construction place, aka:
```c++
    std::vector<int> list = {1,2,3,4,5};

    auto l = list | view::drop(50);    // l is empty() == true NOW
```

**Comment by gnzlbg** [comment]
> Ok, why it becomes empty when I drop exactly all elements in container?

@CaseyCarter when is begin cached, I thought on the first call to begin, but maybe it is cached on construction?

**Comment by CaseyCarter** [comment]
> Is this by the design?

No, this is a bug in `view::drop`'s special case code for random access ranges:

```c++
template<typename Rng, CONCEPT_REQUIRES_(!View<uncvref_t<Rng>>() &&
    std::is_lvalue_reference<Rng>())>
static iterator_range<iterator_t<Rng>, sentinel_t<Rng>>
invoke_(Rng && rng, range_difference_type_t<Rng> n, concepts::RandomAccessRange*)
{
    return {next(begin(rng), n), end(rng)};
}
```

it's trying to be more efficient by simply trimming `n` elements from the left end of the range, but oops `next(begin(rng), n)` will run past the end if the range has less than `n` elements. The fix is to use `next(begin(rng), n, end(rng))` instead, which will stop at `end(rng)` if it gets there before skipping `n` elements.

> The View l is not empty yet. To make it empty you have to "exhaust" it first

This is not true: The iterator returned by `begin` from a `drop` view denotes the first non-dropped element. A `drop` view should be empty if all the elements have been dropped.

> when is begin cached, I thought on the first call to begin

Generally if the adapted view needs an o(1) (I *think* o(f) is the notation for "a function whose scaling is strictly greater than f", it's been a while since I've needed it) calculation to determine its first iterator, that first iterator is cached to provide O(1) for subsequent calls to `begin`. In this specific case the calculation is O(1) at construction time.

**Comment by gnzlbg** [comment]
That makes more sense, thanks @CaseyCarter !

**Comment by CaseyCarter** [comment]
TLDR: `drop` and `drop_exactly` were once a single view that couldn't decide what behavior it wanted for `drop(rng, n)` when `n > size(rng)`. They were split into `drop`, which is an empty range when that condition holds, and `drop_exactly` for which it's a precondition that `n <= size(rng)`. This bug is a vestige of `drop_exactly` that was hiding inside a special case of `drop`.

**Comment by tower120** [comment]
By the way... Why `drop_exactly` exists?...

**Comment by ericniebler** [comment]
> By the way... Why `drop_exactly` exists?...

Performance, and also because it always creates a `SizedRange`, but it assumes to know that the underlying range has enough elements.

**Comment by tower120** [comment]
Hmm... That's was not obvious to me that `drop_exactly` is faster...

Good to know, thanks.
